[agw] [stateless] Add python script to create override config for stateless services

ssanadhya opened the pull request:
## Summary

The current scripts to configure stateless AGW mode edit the Yaml configs in /etc/magma, which is overwritten at the time of software upgrade. This change adds the stateless config as an override Yaml config in /var/opt/magma/configs/ by:

- Adding a new config_stateless_agw.py script to use existing Service Config utils to create override
- Modify sctpd systemd service file to use the new python script for pre and post start steps
- Update S1ap utils to invoke the python script while switching stateless mode on magma-dev VM

**ToDo**: Remove the old shell scripts for config_stateless support as they are not needed any more. This will be part of separate pull request

## Test Plan

- S1AP integration tests
- Ensure that the default config in `/etc/magma/<service_name>.yml` is not modified while running stateless S1ap tests
- Built magma package with `fab dev package:vcs=git` and checked that new python script is included in the package:
```
vagrant@magma-dev:~/magma-packages$ dpkg -c magma_1.3.0-1603173470-13a2cbb7_amd64.deb  | grep config_stateless_agw
-rwxr-xr-x 0/0            5098 2020-10-20 05:58 ./usr/local/bin/config_stateless_agw.py
-rwxr-xr-x 0/0            3110 2020-10-19 16:15 ./usr/local/bin/config_stateless_agw.sh
```
- Tested the newly created Magma and Sctpd package in a physical testbed with real eNB and real UE. Restarting services in stateless mode does not force the UE to re-attach. Override configs are created in `/var/opt/magma/configs/<service_name>.yml`

## Additional Information

[x] This change is backwards breaking
Please destroy and re-provision your magma-dev VM `vagrant provision magma` so that it gets the new Sctpd systemd service file

ssanadhya committed:
Add python script to switch stateful/stateless mode and integrate with
s1ap utils

Signed-off-by: Shruti Sanadhya <ssanadhya@fb.com>

ardzoht commented on lte/gateway/python/scripts/config_stateless_agw.py:
```diff
@@ -0,0 +1,188 @@
+    sys.exit(0)
+
+
+function_map = {
```
define this as uppercase as it will be a constant dict, probably `STATELESS_FUNC_DICT` might be a more descriptive name for this

ardzoht commented on lte/gateway/python/scripts/config_stateless_agw.py:
```diff
@@ -0,0 +1,188 @@
+return_codes = Enum(
+    "return_codes", "STATELESS STATEFUL CORRUPT INVALID", start=0
+)
+service_config_names = [
```
`SERVICE_CONFIG_NAMES`

ardzoht commented on lte/gateway/python/scripts/config_stateless_agw.py:
```diff
@@ -0,0 +1,188 @@
+    for service, config in service_config_names:
+        cfg = load_override_config(service)
+        if cfg is None:
+            cfg = {}
```
`cfg = load_override_config(service) or {}`

ardzoht commented on lte/gateway/deploy/roles/magma/tasks/main.yml:
```diff
@@ -61,7 +61,9 @@
     dest: /usr/local/bin/{{ item }}
     force: yes
     remote_src: yes
+    mode: 0755
```
is it possible to define a permission mode more restrictive for these scripts?

ssanadhya replied:
Removed in latest version.

xjtian commented on lte/gateway/deploy/roles/magma/tasks/main.yml:
```diff
@@ -61,7 +61,9 @@
     remote_src: yes
+    mode: 0755
   with_items:
+    - config_stateless_agw.py
```
chmod this script to match the others so you can get rid of the `mode` key

ssanadhya committed:
Address review comments

Signed-off-by: Shruti Sanadhya <ssanadhya@fb.com>

ssanadhya commented:
@ardzoht @xjtian addressed review comments in https://github.com/magma/magma/pull/2914/commits/b721dd8bd07990f698426b117ee611a9252cb5db . 

The script has same permissions as the existing scripts, I just had to call it through the python3 interpreter.

ssanadhya committed:
Bump up package version for Magma and Sctpd

Signed-off-by: Shruti Sanadhya <ssanadhya@fb.com>

ardzoht reviewed (approved):
lgtm

ssanadhya commented:
RestartIntegrationTests

ulaskozat commented on lte/gateway/python/scripts/config_stateless_agw.py:
```diff
@@ -0,0 +1,186 @@
+def _check_stateless_service_config(service, config_name):
+    config_value = True
+    if service == "pipelined":
+        config_value = False  # pipelined uses inverse logic
```
I would prefer passing expected configuration value for stateless as a named parameter with default value True.

ulaskozat reviewed (approved):
LGTM

ssanadhya committed:
Remove selected keys from Redis on sctpd restart

Signed-off-by: Shruti Sanadhya <ssanadhya@fb.com>

ssanadhya commented:
Merging despite Jenkins CI failing as the failure is not related to this PR and is being investigated separately.

jpad-freedomfi commented:
@ssanadhya I see this was added to apply-v1.3 4 days ago.

I am seeing updates that use to work failing because the sctp.service used to look for .sh script and is now looking for .py script that doesn't exist. 

How is this supposed to be handled for deployed AGWs? 

I see your comment in magma-dev/announce on Oct 28th for people to rebase their dev VMs, how does that apply to a deployed bare metal AGW? 

@berezovskyi-oleksandr Help me here. Any I understanding this change correctly?
